lua_modules/liquidcrystal/liquidcrystal.lua

```diff
@@ -0,0 +1,213 @@
+local LiquidCrystal = {}
+
+-- commands
+local LCD_CLEARDISPLAY = 0x01
+local LCD_RETURNHOME = 0x02
+local LCD_ENTRYMODESET = 0x04
+local LCD_DISPLAYCONTROL = 0x08
+local LCD_CURSORSHIFT = 0x10
+local LCD_FUNCTIONSET = 0x20
+local LCD_SETCGRAMADDR = 0x40
+local LCD_SETDDRAMADDR = 0x80
+
+-- flags for display entry mode
+-- local LCD_ENTRYRIGHT = 0x00
+local LCD_ENTRYLEFT = 0x02
+local LCD_ENTRYSHIFTINCREMENT = 0x01
+-- local LCD_ENTRYSHIFTDECREMENT = 0x00
+
+-- flags for display on/off control
+local LCD_DISPLAYON = 0x04
+-- local LCD_DISPLAYOFF = 0x00
+local LCD_CURSORON = 0x02
+-- local LCD_CURSOROFF = 0x00
+local LCD_BLINKON = 0x01
+-- local LCD_BLINKOFF = 0x00
+
+-- flags for display/cursor shift
+local LCD_DISPLAYMOVE = 0x08
+local LCD_CURSORMOVE = 0x00
+local LCD_MOVERIGHT = 0x04
+local LCD_MOVELEFT = 0x00
+
+-- flags for function set
+local LCD_8BITMODE = 0x10
+local LCD_4BITMODE = 0x00
+local LCD_2LINE = 0x08
+local LCD_1LINE = 0x00
+local LCD_5x10DOTS = 0x04
+local LCD_5x8DOTS = 0x00
+
+-- defaults
+LiquidCrystal._displayfunction = 0
+LiquidCrystal._displaycontrol = 0
+LiquidCrystal._displaymode = 0
+LiquidCrystal._backlight = false
+
+LiquidCrystal.init = function(self, bus, bus_args, fourbitmode, onelinemode, eightdotsmode, column_width, insert_delay)
+   self._bus_args = bus_args
+   if bus == "I2C" then
+      assert(fourbitmode, "8bit mode not supported")
+      self._backend = dofile "i2c4bit.lua"
+   elseif bus == "GPIO" then
+      if fourbitmode then
+	 self._backend = dofile "gpio4bit.lua"
+      else
+	 self._backend = dofile "gpio8bit.lua"
+      end
+   else
+      error(string.format("%s backend is not implemented", bus))
+   end
+
+   if fourbitmode then
+      self._displayfunction = bit.bor(self._displayfunction, LCD_4BITMODE)
+   else
+      self._displayfunction = bit.bor(self._displayfunction, LCD_8BITMODE)
+   end
+
+   if onelinemode then
+      self._displayfunction = bit.bor(self._displayfunction, LCD_1LINE)
+   else
+      self._displayfunction = bit.bor(self._displayfunction, LCD_2LINE)
+   end
+
+   if eightdotsmode then
+      self._displayfunction = bit.bor(self._displayfunction, LCD_5x8DOTS)
+   else
+      self._displayfunction = bit.bor(self._displayfunction, LCD_5x10DOTS)
+   end
+
+   self._offsets = {0, 0x40}
+   if column_width ~= nil then
+      self._offsets[3] =  0 + column_width
+      self._offsets[4] =  0x40 + column_width
+   end
+   self._insert_delay = insert_delay
+
+   self._backend.init(self)
+   self._backend.command(self, bit.bor(LCD_FUNCTIONSET, self._displayfunction))
+   self._backend.command(self, bit.bor(LCD_ENTRYMODESET, self._displaymode))
+
+   self:display(true)
+   self:clear()
+end
+
+LiquidCrystal.clear = function(self)
```

Comment:
A belated style note.  Rather than writing `foo.bar = function(self, args) ... end`, you can write `function foo:bar (args) ... end`; the name `self` is implicit from the `:`; see https://www.lua.org/pil/16.html .

Reply:
Changed function style in `liquidcrystal.lua`. Also, renamed `self` to `screen` in other files.